lib/charms/observability_libs/v1/kubernetes_service_patch.py

```diff
@@ -271,7 +271,11 @@ def is_patched(self) -> bool:
 
     def _is_patched(self, client: Client) -> bool:
         # Get the relevant service from the cluster
-        service = client.get(Service, name=self.service_name, namespace=self._namespace)
+        try:
+            service = client.get(Service, name=self.service_name, namespace=self._namespace)
```

Comment:
ApiError seems a bit too broad to conclude with `False` ([example](https://github.com/canonical/observability-libs/blob/4e20f6ceaa7d51cd1b32aa6eb313fcc7ce371cec/lib/charms/observability_libs/v1/kubernetes_service_patch.py#L248)). Would it make sense to first use `client.list()` to check if the service exists?

Reply:
Good point, I have changed the code a bit to only return `False` when the status is `404` and that we are using a custom name. For all the other cases, I log an error and raise. (Not raising there would crash on the next instruction)

Reply:
Thanks!

The approach we try to take with charm libs is to never to raise, but instead emit a custom event, if necessary (this could be a separate issue if needed). This lib too logs and returns instead of raising ([example](https://github.com/canonical/observability-libs/blob/4e20f6ceaa7d51cd1b32aa6eb313fcc7ce371cec/lib/charms/observability_libs/v1/kubernetes_service_patch.py#L240)).

Is there a particular reason to raise here?

Reply:
The main reason is that the code will crash anyway with another exception if I let it go further because `service` will be `None`. We can move the rest of the method inside the `try` block, but then we will need to return something in the `except` block. It is the main reason I initially returned `False` on all errors. We could however catch the exception up the stack and just log the error.

Reply:
@sed-i Any preference on returning `False` or raising? I do not feel like any of the solutions are perfect, but either way will fix the crash that we are experiencing in some charms.

Reply:
Since we have a public interface,
```python
def is_patched(self) -> bool:
```
I think raising is preferred.

However, when called internally from within the module as part of `def _patch()`, I think we need to absorb the exception to remain consistent:
```diff
-if self._is_patched(client):
-    return

try:
+   if self._is_patched(client):
+       return
    if self.service_name != self._app:
        self._delete_and_create_service(client)
```

Reply:
Makes sense, I applied the change.